### PR TITLE
ROIs and channel names

### PR DESCRIPTION
**Pull request by dominikl**
This PR adds a method to get the channel names of an image and methods to get and add ROIs (Point, Ellipse and Rectangle, supporting rotation transformation).  This would allow proper image segmentation and analysis using the romero-gateway and EBImage package.

E. g. Image segmentation using an example image from idr0021 (just an example, the script is not part of this PR):

|  | dataset_name | image_name                 | roi_index | x                | y                | area | perimeter | radius_min        | radius_max       | radius_mean      | theta                |
|---------|--------------|----------------------------|-----------|------------------|------------------|------|-----------|-------------------|------------------|------------------|----------------------|
| 1       | PCNT-N1      | PCNT_gtub_I_011_SIR_PRJ.dv | 1         | 132.581395348837 | 135.15503875969  | 129  | 33        | 4.83492421809008  | 6.91818911851377 | 5.99228299712261 | -0.839066226573129   |
| 2       | PCNT-N1      | PCNT_gtub_I_011_SIR_PRJ.dv | 2         | 159              | 135.285714285714 | 7    | 6         | 0.666666666666657 | 1.33333333333334 | 1.08531432618137 | 4.45652403883732e-11 |
| 3       | PCNT-N1      | PCNT_gtub_I_011_SIR_PRJ.dv | 3         | 139.769230769231 | 177.676923076923 | 130  | 33        | 5.29367139527366  | 6.82632510336976 | 6.03794953134715 | -1.28545723486094    |

screen shot 2018-10-31 at 10 24 26

**Also bumps the OMERO version to the latest 5.4.9**

/cc @jburel 

# Test

Try the ROIs.R example (last commit)

### CODE REVIEW

**Comment by jburel** [comment]
Did you look at adding polyline support?

**Comment by dominikl** [comment]
Was mostly focussing on what I can get from EBImage tools (x, y, radius, etc.), haven't found how I could get polyline / polygon coordinates yet. That's why I've not included polyline support yet, but can do so of course. EBImage segmentation creates masks, haven't found out how I could get polygons or polylines from that.

**Comment by jburel** [comment]
saving masks will be useful too 
It won't be supported in figure, but iviewer should work

**Comment by dominikl** [comment]
Ah, that's great. I'll try to add that then.

**Comment by dominikl** [comment]
Do we have an example of a "mask" somewhere? I'm not sure what iviewer supposed to show. I tried to create a view masks which cover the whole image (x=0, y=0, w=image_width, h=image_height) with content 0's for the background and 1's (or 255's) for the ROIs. iviewer shows the outline of the mask (which is the whole image) but nothing else.

**Comment by dominikl** [comment]
I'll add the Mask ROI support in another PR. This will need some additions in the 'MaskData' Java gateway class, too, as this class is very difficult to use at the moment.

**Comment by mtbc** [comment]
Some of the shape geometry's a bit hairy, like in adding rotated ROIs, don't know if we'd want some extra tests around that.

File: examples/ROIs.R [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+rois <- data.frame(x = c(0), y = c(0), rx = c(0), ry = c(0), w = c(0), h = c(0), theta = c(0), text = c('remove'), stringsAsFactors = FALSE)
+rois <- rbind(rois, c(10, 10, NA, NA, NA, NA, NA, "This is a point"))
+rois <- rbind(rois, c(10, 10, 10, 20, NA, NA, NA, "This is an ellipse"))
+rois <- rbind(rois, c(10, 10, NA, NA, 10, 20, NA, "This is a Rectangle"))
+rois <- rbind(rois, c(10, 10, NA, NA, 10, 20, 0.5, "This is a rotated Rectangle"))
+rois <- rois[-1,]
+rownames(rois) <- c()
+
+# Add the ROIs to the image
+addROIs(img, coords = rois)
```
Comment (jburel) [inline examples/ROIs.R]: This should be ``image``

File: examples/ROIs.R [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+# Get the ROIs again
+rois <- getROIs(img)
```
Comment (jburel) [inline examples/ROIs.R]: same

Comment (jburel) [reply examples/ROIs.R]: Invoking ``addROIs(image, coords = rois)`` prints out ``NULL``
This can be confusing

**Comment by jburel** [comment]
centrin_pcnt_cep215_20110506_fri-1545_0_sir_prj dv
ROIs added and visible in iviewer

**Comment by jburel** [comment]
The output of ``getChannelNames(image)`` is ``"CENT2"      "PCNT"       "CDK5RAP2-C"`` as expected

File: R/Image.R [hunk 3]
```diff
@@ -98,6 +153,261 @@ setMethod(
...
+            ry <- as.numeric(ry)
+            text <- tryCatch( {
+              p$getText()},
+              error=function(err) {
+                return(NA)
+              })
+            text <- as.character(text)
+            rois <- rbind(rois, c(x, y, rx, ry, NA, NA, th, text))
+          }
+          if(.jinstanceof(shape, RectangleData)) {
```
Comment (jburel) [inline R/Image.R]: space after ``if``

File: R/Image.R [hunk 4]
```diff
@@ -98,6 +153,261 @@ setMethod(
...
+            y <- as.numeric(y)
+            text <- tryCatch( {
+              p$getText()},
+              error=function(err) {
+                return(NA)
+              })
+            text <- as.character(text)
+            rois <- rbind(rois, c(x, y, NA, NA, NA, NA, th, text))
+          }
+          if(.jinstanceof(shape, EllipseData)) {
```
Comment (jburel) [inline R/Image.R]: space

File: R/Image.R [hunk 5]
```diff
@@ -98,6 +153,261 @@ setMethod(
...
+          trans <- shape$getTransform()
+          th <- NA
+          if (!is.null(trans)) {
+            # Assumption: The affine transform is a rotation about
+            # an angle theta
+            th <- trans$getA00()$getValue()
+            th <- acos(as.numeric(th))
+          }
+          
+          if(.jinstanceof(shape, PointData)) {
```
Comment (jburel) [inline R/Image.R]: space

File: R/Image.R [hunk 6]
```diff
@@ -98,6 +153,261 @@ setMethod(
...
+    it <- roiresults$iterator()
+    while(it$hasNext()) {
+      roiresult <- .jrcall(it, method = "next")
+      roidatas <- roiresult$getROIs()
+      it2 <- roidatas$iterator()
+      while(it2$hasNext()) {
+        roidata <- .jrcall(it2, method = "next")
+        shapes <- roidata$getShapes(z, t)
+        it3 <- shapes$iterator()
+        while(it3$hasNext()) {
```
Comment (jburel) [inline R/Image.R]: space after ``while``

File: R/Image.R [hunk 7]
```diff
@@ -98,6 +153,261 @@ setMethod(
...
+    t <- as.integer(t - 1)
+    
+    roiresults <- fac$loadROIsByPlane(ctx, iid, z, t)
+    rois <- data.frame(x = c(0), y = c(0), rx = c(0), ry = c(0), w = c(0), h = c(0), theta = c(0), text = c('remove'), stringsAsFactors = FALSE)
+    it <- roiresults$iterator()
+    while(it$hasNext()) {
+      roiresult <- .jrcall(it, method = "next")
+      roidatas <- roiresult$getROIs()
+      it2 <- roidatas$iterator()
+      while(it2$hasNext()) {
```
Comment (jburel) [inline R/Image.R]: space

File: R/Image.R [hunk 8]
```diff
@@ -98,6 +153,261 @@ setMethod(
...
+    iid <- obj$getId()
+    iid <- .jlong(iid)
+    
+    z <- as.integer(z - 1)
+    t <- as.integer(t - 1)
+    
+    roiresults <- fac$loadROIsByPlane(ctx, iid, z, t)
+    rois <- data.frame(x = c(0), y = c(0), rx = c(0), ry = c(0), w = c(0), h = c(0), theta = c(0), text = c('remove'), stringsAsFactors = FALSE)
+    it <- roiresults$iterator()
+    while(it$hasNext()) {
```
Comment (jburel) [inline R/Image.R]: space

File: R/Image.R [hunk 9]
```diff
@@ -98,6 +153,261 @@ setMethod(
...
+    obj <- image@dataobject
+    gateway <- getGateway(server)
+    ctx <- getContext(server)
+    
+    fac <- gateway$getFacility(MetadataFacility$class)
+    iid <- image@dataobject$getId()
+    jchannels <- fac$getChannelData(ctx, .jlong(iid))
+    channels <- c()
+    it <- jchannels$iterator()
+    while(it$hasNext()) {
```
Comment (jburel) [inline R/Image.R]: space after ``while``

File: R/Image.R [hunk 10]
```diff
@@ -46,6 +46,61 @@ setGeneric(
...
+#' Get the ROIs (Point, Ellipse or Rectangle) of an Image
+#' x, y -> Point
+#' x, y, rx, ry -> Ellipse
+#' x, y, w, h -> Rectangle
+#' theta: Rotation about theta
+#' text: An optional comment
+#' @param image The image
+#' @param z The Z plane (default: 1)
+#' @param t The timepoint (default: 1)
+#' @return The coordinates as dataframe (x, y, rx, ry, w, h, theta, text)
```
[the remark this answers is not in the document]
Comment (dominikl) [reply R/Image.R]: I don't know... I think polygons need a different method (like mask ROIs will do).

**Comment by jburel** [comment]
No more "NULL" printed out

File: examples/ROIs.R [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+server <- connect(server)
+
+# Load the image
+imageId <- 1
+image <- loadObject(server, "ImageData", imageId)
+
+# Specify some ROIs as dataframe
+rois <- data.frame(x = c(0), y = c(0), rx = c(0), ry = c(0), w = c(0), h = c(0), theta = c(0), text = c('remove'), stringsAsFactors = FALSE)
+rois <- rbind(rois, c(10, 10, NA, NA, NA, NA, NA, "This is a point"))
+rois <- rbind(rois, c(10, 10, 10, 20, NA, NA, NA, "This is an ellipse"))
```
[the remark this answers is not in the document]
Comment (jburel) [reply examples/ROIs.R]: Discussed with @dominikl, this was not intended but we can adjust the example in a follow-up
PR
Merging so we can tag ahead of the next training

**Pull request merged**